### ISSUE TITLE
Null check with 'as' does not update nullable state

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#38522]
**Version Used**: 16.3 Preview 2

**Steps to Reproduce**:

```csharp
SomeType? value = null;
if (value?.Property as AnotherType != null)
{
  _ = value.Property;
}

SomeType? value2 = null;
if (value2?.Property is AnotherType)
{
  _ = value2.Property;
}
```

**Expected Behavior**:

No warning is reported on the access of `value.Property` or `value2.Property`.

**Actual Behavior**:

A warning is reported on the access of `value.Property`, but no warning is reported on the access of 'value2.Property`.

**Comment by jaredpar** [comment]
Here is a simpler case which demonstrates the underlying behavior here:

``` csharp
using System;
#nullable enable
public class C
{
    public void M(object? o)
    {
    	if (o as string != null)
        {
        	o.ToString(); // Warning
        }
        
        if (o is string)
        {
        	o.ToString();      
        }
    }
}
```

**Comment by jaredpar** [comment]
At the moment we don't track state across the `as` operator even when in the presence of an explicit `null` check. The behavior is intentional at the moment but is an area we could consider expanding to support if we see enough cases of this.

**Comment by RikkiGibson** [comment]
Did we find places in Roslyn where this introduced unnecessary warnings?

**Comment by jaredpar** [comment]
@RikkiGibson 

> Did we find places in Roslyn where this introduced unnecessary warnings?

Not that I'm aware of. I did a quick search of the code base using the regex `as (\w+) !=` and only got a single hit that was in a comment.
